onnxruntime/core/providers/webgpu/nn/layer_norm.cc

```diff
@@ -0,0 +1,170 @@
+
+namespace onnxruntime {
+namespace webgpu {
+
+static uint32_t getMaxComponents(int size) {
+  if (size % 4 == 0) {
+    return 4;
+  } else if (size % 2 == 0) {
+    return 2;
+  }
+  return 1;
+}
+
+static int normalizeAxis(int axis, int tensorRank) {
+  if (axis < -tensorRank && axis >= tensorRank) {
+    ORT_THROW("invalid axis: ", axis);
+  }
+  return axis < 0 ? axis + tensorRank : axis;
+}
+
+static std::string fillVar(std::string dataType, int components, std::string value) {
+  if (components == 1) {
+    return dataType + "(" + value + ")";
+  }
+  return "vec" + std::to_string(components) + "<" + dataType + ">(" + value + ")";
+}
+
+static std::string castToF32(int components, std::string value) {
+  if (components == 1) {
+    return "f32(" + value + ")";
+  }
+  return "vec" + std::to_string(components) + "<f32>(" + value + ")";
+};
+
+static std::string sumVector(std::string x, int components) {
+  switch (components) {
+    case 1:
+      return x;
+    case 2:
+      return "(" + x + ".x + " + x + ".y" + ")";
+    case 4:
+      return "(" + x + ".x + " + x + ".y + " + x + ".w + " + x + ".z" + ")";
+    default:
+      ORT_THROW("Unsupported number of components: ", components);
+  }
+}
+
+Status LayerNormProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const auto& x = shader.AddInput("x", ShaderUsage::UseUniform | ShaderUsage::UseValueTypeAlias);
+  shader.AddInput("scale", ShaderUsage::UseUniform);
+  if (has_bias_) {
+    shader.AddInput("bias", ShaderUsage::UseUniform);
+  }
+  shader.AddOutput("output", ShaderUsage::UseUniform);
+
+  int components = x.NumComponents();
+  std::string bias = (has_bias_) ? " + bias[j] " : "";
+  std::string simpl1 = (simplified_) ? "" : "- mean * mean ";
+  std::string simpl2 = (simplified_) ? "" : "- mean ";
+  std::string fillvec = fillVar("f32", components, "0");
+  std::string element_type = (isFP16_) ? "f16;\n" : "f32;\n";
+
+  shader.AppendImplementation("alias element_t = " + element_type);
+
+  std::stringstream ss;
+  ss << "let offset = global_idx * uniforms.norm_size_vectorized;\n"
+     << "var mean_vector = " << fillvec << ";\n"
+     << "var mean_square_vector = " << fillvec << ";\n"
+     << "for (var h: u32 = 0u; h < uniforms.norm_size_vectorized; h++) {\n"
+     << "   let value = " << castToF32(components, "x[h + offset]") << ";\n"
+     << "   mean_vector += value;\n"
+     << "   mean_square_vector += value * value;\n"
+     << "}\n"
+     << "let mean = " << sumVector("mean_vector", components) << " / f32(uniforms.norm_size);\n"
+     << "let inv_std_dev = inverseSqrt(" << sumVector("mean_square_vector", components) << " / f32(uniforms.norm_size) " << simpl1 << "+ uniforms.epsilon);\n"
+     << "for (var j: u32 = 0; j < uniforms.norm_size_vectorized; j++) {\n"
+     << "   let f32input = " << castToF32(components, "x[j + offset]") << ";\n"
+     << "   let f32scale = " << castToF32(components, "scale[j]") << ";\n"
```

Comment:
can use `alias f32_compute_t` to `f32` or `vecN<f32>` (depending on component_num), so can use `f32_compute_t(scale[j])` here